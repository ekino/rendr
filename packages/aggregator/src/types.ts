import { BlockDefinition, Page, RendrCtx } from "@ekino/rendr-core";

export type HandlerList = {
  [index: string]: Handler;
};

export type Handler = (
  definition: BlockDefinition,
  ctx: RendrCtx,
  page: Page
) => Promise<BlockDefinition>;

export type HandlerRegistry = (code: string) => Handler;
