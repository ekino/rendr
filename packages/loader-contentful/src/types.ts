import { Asset as ContentfulAsset, Entry } from "contentful";
import { RequestCtx } from "@ekino/rendr-core";

export { Asset as ContentfulAsset } from "contentful";

import { ContentfulClientApi } from "contentful";

export interface EntryNormalizer {
  (entry: Entry<any> | ContentfulAsset, normalizers?: EntryNormalizer):
    | any
    | void;
}

export type EntryNormalizerList = {
  [index: string]: EntryNormalizer;
};

export interface ClientFactory {
  (ctx: RequestCtx): ContentfulClientApi;
}

export interface Website {
  id: string;
  name: string;
  domains: string[];
  path: string;
  culture: string;
  countryCode: string;
  order: number;
  enabled: boolean;
}

export interface Asset {
  url: string;
  title: string;
  id: string;
}
export interface ContentfulWebsite {
  name: string;
  domains: string[];
  path: string;
  culture: string;
  country_code: string;
  order: number;
  enabled: boolean;
}

export interface ContentfulPage {
  title: string;
  type: string;
  seo_description: string;
  seo_keywords: string;
  extends: string;
  path: string;
  website: ContentfulWebsite;
  layout: string;
  ttl: number;
  settings: object;
  published_at: string;
  blocks: Entry<any>[];
  code: string;
}

export interface BaseContentfulBlock {
  internal_title: string;
  container: string;
  order: number;
}
