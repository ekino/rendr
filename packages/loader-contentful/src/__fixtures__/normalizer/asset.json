{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "8VoLF5Jla5JcMnhNV74Oj",
        "type": "Asset",
        "createdAt": "2019-08-09T21:30:18.823Z",
        "updatedAt": "2019-08-09T21:30:18.823Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "locale": "en-US"
    },
    "fields": {
        "title": "test",
        "description": "test",
        "file": {
            "url": "//images.ctfassets.net/f5bx295b472q/8VoLF5Jla5JcMnhNV74Oj/4f3afcf599ffbcc55528e2026e40cd0d/Screenshot_2019-07-18_at_16.39.23.png",
            "details": {
                "size": 10072,
                "image": {
                    "width": 210,
                    "height": 134
                }
            },
            "fileName": "Screenshot 2019-07-18 at 16.39.23.png",
            "contentType": "image/png"
        }
    }
}