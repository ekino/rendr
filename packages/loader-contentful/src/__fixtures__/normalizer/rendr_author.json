{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "7pOMMe1TEojuNC9gKdxR0k",
        "type": "Entry",
        "createdAt": "2019-08-20T18:39:11.532Z",
        "updatedAt": "2019-08-20T18:39:11.532Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_author"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "name": "Thomas Rabaix",
        "job_title": "CTO",
        "slug": "thomas-rabaix",
        "social_twitter": "th0masr",
        "biography": "CTO at Ekino. Sonata Project founder. Symfony Expert. Lately doing python, go, ansible and cloud platforms."
    }
}