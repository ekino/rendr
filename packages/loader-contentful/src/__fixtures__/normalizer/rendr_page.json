{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "4QrVSXjiXz6iSS9Vli7sDh",
        "type": "Entry",
        "createdAt": "2019-08-20T18:26:10.936Z",
        "updatedAt": "2019-08-20T18:28:43.549Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 3,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_page"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "title": "Ekino Homepage",
        "seo_description": "Homepage",
        "seo_keywords": "digital, awesome",
        "extends": "root",
        "path": "/",
        "blocks": [
            {
                "sys": {
                    "space": {
                        "sys": {
                            "type": "Link",
                            "linkType": "Space",
                            "id": "f5bx295b472q"
                        }
                    },
                    "id": "1ovjOotrujpd5BTTKIxq8C",
                    "type": "Entry",
                    "createdAt": "2019-08-20T18:25:58.971Z",
                    "updatedAt": "2019-08-20T18:25:58.971Z",
                    "environment": {
                        "sys": {
                            "id": "master",
                            "type": "Link",
                            "linkType": "Environment"
                        }
                    },
                    "revision": 1,
                    "contentType": {
                        "sys": {
                            "type": "Link",
                            "linkType": "ContentType",
                            "id": "rendr_block_text"
                        }
                    },
                    "locale": "en-US"
                },
                "fields": {
                    "internal_title": "About section",
                    "container": "header",
                    "order": 1,
                    "title": "About",
                    "contents": "We create digital services that transform businesses.\n\nUsing lean, Agile methodologies and rapid prototyping, it’s never been easier to create, test, and refine new services based on customer data and human behaviour.\n\nTogether, we’ll collaborate to accelerate your innovation and increase your differentiation."
                }
            }
        ],
        "layout": "default",
        "published_at": "2019-08-16T00:00",
        "settings": {
            "foo": "bar"
        }
    }
}