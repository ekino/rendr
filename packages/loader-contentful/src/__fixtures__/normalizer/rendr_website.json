{
    "sys": {
        "space": {
            "sys": {
                "type": "Link",
                "linkType": "Space",
                "id": "f5bx295b472q"
            }
        },
        "id": "75Vqk3IpfxlzQfvhdSASZV",
        "type": "Entry",
        "createdAt": "2019-08-20T18:50:08.851Z",
        "updatedAt": "2019-08-20T18:50:08.851Z",
        "environment": {
            "sys": {
                "id": "master",
                "type": "Link",
                "linkType": "Environment"
            }
        },
        "revision": 1,
        "contentType": {
            "sys": {
                "type": "Link",
                "linkType": "ContentType",
                "id": "rendr_website"
            }
        },
        "locale": "en-US"
    },
    "fields": {
        "name": "Ekino.com",
        "domains": [
            "www.ekino.com"
        ],
        "path": "/",
        "culture": "en_GB",
        "country_code": "FRA",
        "order": 1,
        "enabled": true
    }
}