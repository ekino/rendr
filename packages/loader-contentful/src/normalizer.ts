import { Entry } from "contentful";
import {
  Page,
  Map,
  BlockDefinition,
  Settings,
  NormalizationError,
  createPage,
  RendrCtx,
} from "@ekino/rendr-core";

import {
  Asset,
  Website,
  ContentfulWebsite,
  ContentfulPage,
  ContentfulAsset,
  EntryNormalizerList,
  EntryNormalizer,
  BaseContentfulBlock,
} from "./types";

const ASSET_REGEXP = /(images\.(ctfassets\.net|contentful\.com)([0-9a-zA-Z/\-_]*)\.([a-z]{1,4}))/g;
const ASSET_FORMATS: Map = {
  jpg: "jpg",
  jpeg: "jpg",
  png: "png",
};

export const emptyPicture = {
  enabled: false,
  id: "-",
  title: "No picture",
  url: "",
};

export const defaultNormalizers: EntryNormalizerList = {
  rendr_page: normalizePage,
  rendr_website: normalizeWebsite,
  asset: normalizePicture, // Asset is the default from contentul
};

export function createNormalizer(
  extraNormalizers: EntryNormalizerList = {}
): EntryNormalizer {
  const normalizers: EntryNormalizerList = {
    ...defaultNormalizers,
    ...extraNormalizers,
  };

  // check for recursive loop, and protect against
  // infinite loop.
  const stack: string[] = [];

  return async function normalizer(
    ctx: RendrCtx,
    entry: Entry<any> | ContentfulAsset,
    normalizerFn?: EntryNormalizer
  ): Promise<any> {
    if (!validEntry(entry)) {
      return;
    }

    if (stack.includes(entry.sys.id)) {
      // recursive call cannot normalize
      return false;
    }

    let key;
    if (entry.sys.type == "Asset") {
      key = "asset";
    } else if (
      entry.sys.type == "Entry" &&
      entry.sys.contentType.sys.id in normalizers
    ) {
      key = entry.sys.contentType.sys.id;
    } else {
      // entry.sys.type == 'Link' if an entity is not resolvable
      // ie: Contentful API returns a `notResolvable` error.
      return;
    }

    stack.push(entry.sys.id);

    try {
      return await normalizers[key](ctx, entry, normalizer);
    } catch (err) {
      // log error
      if (err instanceof NormalizationError) {
        throw err;
      }

      throw new NormalizationError(err);
    } finally {
      stack.pop();
    }
  };
}

export function fixImageUrl(content: string) {
  if (!content) {
    return "";
  }

  return content.replace(ASSET_REGEXP, (replacement, ...params: string[]) => {
    if (params[3] in ASSET_FORMATS) {
      return `images.ctfassets.net${params[2]}.${params[3]}?fm=${
        ASSET_FORMATS[params[3]]
      }`;
    }

    return replacement;
  });
}

export function fixHttps(url: string) {
  if (url.length < 2) {
    return url;
  }

  if (url.substr(0, 2) == "//") {
    return `https:${url}`;
  }

  return url;
}

export function validEntry(
  entity: Entry<any> | ContentfulAsset,
  contentType = ""
) {
  if (!entity) {
    return false;
  }

  if (!("sys" in entity) || !("id" in entity.sys)) {
    return false;
  }

  if (contentType.length === 0) {
    return true;
  }

  return entity.sys.contentType.sys.id === contentType;
}

export async function normalizePage(
  ctx: RendrCtx,
  entry: Entry<ContentfulPage>,
  normalizer: EntryNormalizer
): Promise<Page> {
  const ttl = ctx.settings.rendr_site
    ? (ctx.settings.rendr_site as Website).cache.ttl
    : 0;
  const sharedTtl = ctx.settings.rendr_site
    ? (ctx.settings.rendr_site as Website).cache.sharedTtl
    : 0;

  const data = {
    id: entry.sys.id,
    head: {
      title: entry.fields.title,
      meta: [] as Settings[],
    },
    path: entry.fields.path ?? "",
    template: entry.fields.layout ?? "default",
    settings: entry.fields.settings,
    blocks: [] as BlockDefinition[],
    cache: {
      ttl: entry.fields.ttl ?? ttl,
      sharedTtl: entry.fields.shared_ttl ?? sharedTtl,
    },
  };

  const keys = Object.keys(entry.fields);

  for (let k in keys) {
    const key = keys[k];

    // NAME_blocks convention
    if (key.substr(-7) !== "_blocks") {
      continue;
    }

    // @ts-ignore
    const blocks = entry.fields[key] as Entry<any>[];
    const container = ((word: string) => {
      const words = word.split("_blocks");
      words.pop();

      return words.join();
    })(key);

    if (!blocks) {
      continue;
    }

    for (let index in blocks) {
      const block = blocks[index];

      const def = await normalizer(ctx, block);

      if (!def) {
        continue;
      }

      def.id = block.sys.id;
      def.container = container;
      def.order = index;

      data.blocks.push(def);
    }
  }

  if (entry.fields.seo_keywords && entry.fields.seo_keywords.length > 0) {
    data.head.meta.push({
      name: "keywords",
      content: entry.fields.seo_keywords,
    });
  }

  if (entry.fields.seo_description && entry.fields.seo_description.length > 0) {
    data.head.meta.push({
      name: "description",
      content: entry.fields.seo_description,
    });
  }

  return createPage(data);
}

export function normalizePicture(ctx: RendrCtx, entry: ContentfulAsset): Asset {
  return {
    id: entry.sys.id,
    title: entry.fields.title,
    url: fixHttps(fixImageUrl(entry.fields.file.url)),
  };
}

export function normalizeWebsite(
  ctx: RendrCtx,
  site: Entry<ContentfulWebsite>,
  normalizers: EntryNormalizer
): Website {
  return {
    id: site.sys.id,
    name: site.fields.name,
    path: site.fields.path,
    domains: site.fields.domains,
    culture: site.fields.culture,
    countryCode: site.fields.country_code,
    order: site.fields.order ? site.fields.order : 99,
    enabled: !!site.fields.enabled,
    cache: {
      sharedTtl: site.fields.shared_ttl ?? 0,
      ttl: site.fields.ttl ?? 0,
    },
    mainMenu: site.fields.main_menu ? site.fields.main_menu : {},
    settings: site.fields.settings ? site.fields.settings : {},
  };
}

export function createBlockDefinition(
  entry: Entry<BaseContentfulBlock>,
  type: string,
  settings: Settings
): BlockDefinition {
  return {
    id: entry.sys.id,
    container: "body", // default value for now
    order: 99, // default value for now
    settings: settings,
    type: type,
    meta: {},
  };
}
