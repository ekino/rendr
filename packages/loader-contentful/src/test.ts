import { Entry, ContentfulClientApi } from "contentful";
import { readFileSync } from "fs";

export function createDummyEntry<T>(data: T): Entry<T> {
  return {
    sys: {
      type: "Entry",
      id: "",
      createdAt: "",
      updatedAt: "",
      locale: "fr_FR",
      contentType: {
        sys: {
          type: "Link",
          linkType: "ContentType",
          id: "DummyEntry",
        },
      },
      space: {
        sys: {
          type: "Link",
          linkType: "Space",
          id: "DummySpace",
        },
      },
    },
    fields: data,
    toPlainObject: () => ({}),
    update: () => Promise.resolve(this),
  };
}

export function loadJson(path: string) {
  return JSON.parse(readFileSync(path, { encoding: "utf8" }));
}

export function getMockClient(
  methods: Partial<ContentfulClientApi>
): ContentfulClientApi {
  return {
    getAsset: async (id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getAssets: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getContentType: async (id: string) => {
      throw new Error("The method should not be called");
    },
    getContentTypes: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntries: async <T>(query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntry: async <T>(id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getSpace: async () => {
      throw new Error("The method should not be called");
    },
    getLocales: async () => {
      throw new Error("The method should not be called");
    },
    parseEntries: async <T>(raw: any) => {
      throw new Error("The method should not be called");
    },
    sync: async (query: any) => {
      throw new Error("The method should not be called");
    },
    ...methods,
  };
}
