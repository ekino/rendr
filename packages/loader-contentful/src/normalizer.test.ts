import { fixImageUrl, validEntry } from "./normalizer";
import { createDummyEntry } from "./test";
import { createNormalizer } from "./index";
import { EntryNormalizerList } from "./types";
import { Entry } from "contentful";
import { loadJson, createContext } from "./test";
import { RequestCtx } from "@ekino/rendr-core";

interface Wheel {
  brand: string;
  size: number;
}

interface Car {
  speed: number;
  name: string;
  data: number; // the normalization will update this value
  wheel: Wheel; // the normalization will update this value
}

interface ContentfulWeel {
  info: string;
}

interface ContentfulCar {
  speed: number;
  name: string;
  data: string;
  wheel: Entry<ContentfulWeel>;
}

describe("test fix url", () => {
  it("should fix single url", () => {
    const contents = [
      {
        source:
          "//images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.jpg",
        expected:
          "//images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.jpg?fm=jpg",
      },
      {
        source:
          "//images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.png hello //images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.jpeg",
        expected:
          "//images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.png?fm=png hello //images.ctfassets.net/95wnqgvmhlea/2IPecD9vXywcucqKWiQu8S/3e6cf50875e99bb6230cd7dc54553206/anne-marte.jpeg?fm=jpg",
      },
      {
        source:
          '<a href="//images.contentful.com/95wnqgvmhlea/2lffMyV2ccEuI2cMkqyigs/367f93360c80948ae3e208662724e14a/run.png"><img alt="Figure 3. Screenshot illustrating the way to run a jupyter notebook cell." src="//images.contentful.com/95wnqgvmhlea/2lffMyV2ccEuI2cMkqyigs/367f93360c80948ae3e208662724e14a/run.png"></a>',
        expected:
          '<a href="//images.ctfassets.net/95wnqgvmhlea/2lffMyV2ccEuI2cMkqyigs/367f93360c80948ae3e208662724e14a/run.png?fm=png"><img alt="Figure 3. Screenshot illustrating the way to run a jupyter notebook cell." src="//images.ctfassets.net/95wnqgvmhlea/2lffMyV2ccEuI2cMkqyigs/367f93360c80948ae3e208662724e14a/run.png?fm=png"></a>',
      },
    ];

    contents.forEach((v) => {
      expect(fixImageUrl(v.source)).toEqual(v.expected);
    });
  });
});

describe("test validEntry", () => {
  it("with invalid entry", () => {
    // @ts-ignore
    expect(validEntry(false)).toBeFalsy();

    // @ts-ignore
    expect(validEntry({})).toBeFalsy();
  });

  it("with a valid entry but no contentType", () => {
    const entry = createDummyEntry<any>({});

    expect(validEntry(entry)).toBeTruthy();
  });

  it("with a valid entry but with specific invalid contentType", () => {
    const entry = createDummyEntry<any>({});

    expect(validEntry(entry, "FooBar")).toBeFalsy();
  });

  it("with a valid entry but with specific valid contentType", () => {
    const entry = createDummyEntry<any>({});

    expect(validEntry(entry, "DummyEntry")).toBeTruthy();
  });
});

describe("test normalizer", () => {
  it("with default values (no normalizer)", () => {
    const normalizer = createNormalizer();

    const ctx = createContext({});
    const entry = createDummyEntry<ContentfulCar>({
      speed: 10,
      name: "2CV",
      data: "1",
      wheel: createDummyEntry<ContentfulWeel>({
        info: "",
      }),
    });

    const result = normalizer(ctx, entry);

    expect(result).toBeUndefined();
  });

  it("with default values with normalizer", () => {
    const normalizers: EntryNormalizerList = {
      car: (ctx: RequestCtx, entry: Entry<ContentfulCar>, normalizer) => {
        return {
          speed: entry.fields.speed,
          name: entry.fields.name,
          data: parseInt(entry.fields.data, 10),
          wheel: normalizer(ctx, entry.fields.wheel),
        };
      },
      wheel: (ctx: RequestCtx, entry: Entry<ContentfulWeel>) => {
        const split = entry.fields.info.split(",");

        return {
          brand: split[0],
          size: parseInt(split[1], 10),
        };
      },
    };
    const normalizer = createNormalizer(normalizers);

    const ctx = createContext({});
    const entry = createDummyEntry<ContentfulCar>({
      speed: 10,
      name: "2CV",
      data: "1",
      wheel: createDummyEntry<ContentfulWeel>({
        info: "michelin,17",
      }),
    });
    entry.sys.contentType.sys.id = "car"; // so the normalizer can catch the value
    entry.fields.wheel.sys.contentType.sys.id = "wheel";

    const result = normalizer(ctx, entry);

    expect(result).toMatchSnapshot();
  });

  const files = [
    "rendr_page",
    "rendr_page_missing_field",
    "rendr_website",
    "asset",
  ];

  const ctx = createContext({});

  files.forEach((file) => {
    it(`test ${file}.json`, () => {
      const normalizer = createNormalizer({});
      const entry = loadJson(
        `${__dirname}/__fixtures__/normalizer/${file}.json`
      );
      const block = normalizer(ctx, entry);

      expect(block).toBeDefined();
      expect(block).toMatchSnapshot();
    });
  });
});
