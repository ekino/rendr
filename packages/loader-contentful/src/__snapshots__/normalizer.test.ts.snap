// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test normalizer test asset.json 1`] = `
Object {
  "id": "8VoLF5Jla5JcMnhNV74Oj",
  "title": "test",
  "url": "https://images.ctfassets.net/f5bx295b472q/8VoLF5Jla5JcMnhNV74Oj/4f3afcf599ffbcc55528e2026e40cd0d/Screenshot_2019-07-18_at_16.39.23.png",
}
`;

exports[`test normalizer test rendr_page.json 1`] = `
Page {
  "blocks": Array [],
  "cache": Object {
    "ttl": 0,
  },
  "head": Object {
    "defaultTitle": "-",
    "htmlAttributes": Object {},
    "link": "",
    "meta": Array [
      Object {
        "content": "digital, awesome",
        "name": "keywords",
      },
      Object {
        "content": "Homepage",
        "name": "description",
      },
    ],
    "title": "Ekino Homepage",
    "titleTemplate": "%s",
  },
  "id": "4QrVSXjiXz6iSS9Vli7sDh",
  "path": "/",
  "settings": Object {
    "foo": "bar",
  },
  "statusCode": 200,
  "template": "default",
  "type": "document",
}
`;

exports[`test normalizer test rendr_page_missing_field.json 1`] = `
Page {
  "blocks": Array [],
  "cache": Object {
    "ttl": 0,
  },
  "head": Object {
    "defaultTitle": "-",
    "htmlAttributes": Object {},
    "link": "",
    "meta": Array [
      Object {
        "content": "digital, awesome",
        "name": "keywords",
      },
      Object {
        "content": "Homepage",
        "name": "description",
      },
    ],
    "title": "Ekino Homepage",
    "titleTemplate": "%s",
  },
  "id": "4QrVSXjiXz6iSS9Vli7sDh",
  "path": "/",
  "settings": Object {},
  "statusCode": 200,
  "template": "default",
  "type": "document",
}
`;

exports[`test normalizer test rendr_website.json 1`] = `
Object {
  "countryCode": "FRA",
  "culture": "en_GB",
  "domains": Array [
    "www.ekino.com",
  ],
  "enabled": true,
  "id": "75Vqk3IpfxlzQfvhdSASZV",
  "name": "Ekino.com",
  "order": 1,
  "path": "/",
}
`;

exports[`test normalizer with default values with normalizer 1`] = `
Object {
  "data": 1,
  "name": "2CV",
  "speed": 10,
  "wheel": Object {
    "brand": "michelin",
    "size": 17,
  },
}
`;
